expression/builtin_op.go

```diff
@@ -378,6 +394,165 @@ func (b *builtinUnaryOpSig) eval(row []types.Datum) (d types.Datum, err error) {
 	return
 }
 
+type unaryMinusFunctionClass struct {
+	baseFunctionClass
+}
+
+func (b *unaryMinusFunctionClass) typeInfer(argExpr Expression) (evalTp, bool) {
+	tp := tpInt
+	switch argExpr.GetType().Tp {
+	case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeFloat,
+		mysql.TypeDatetime, mysql.TypeDuration, mysql.TypeTimestamp:
+		tp = tpReal
+	case mysql.TypeNewDecimal:
+		tp = tpDecimal
+	}
+
+	overflow := false
+	// TODO: handle float overflow
+	if arg, ok := argExpr.(*Constant); ok {
+		switch arg.Value.Kind() {
+		case types.KindUint64:
+			uval := arg.Value.GetUint64()
+			if uval > uint64(-math.MinInt64) {
+				overflow = true
+				tp = tpDecimal
+			}
+		case types.KindInt64:
+			val := arg.Value.GetInt64()
+			if val == math.MinInt64 {
+				overflow = true
+				tp = tpDecimal
+			}
+		}
+	}
+	return tp, overflow
+}
+
+func (b *unaryMinusFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	argExpr := args[0]
+	retTp, intOverflow := b.typeInfer(argExpr)
+
+	switch argExpr.GetTypeClass() {
+	case types.ClassInt:
+		if intOverflow {
+			bf, err := newBaseBuiltinFuncWithTp(args, ctx, retTp, tpDecimal)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			sig = &builtinUnaryMinusDecimalSig{baseDecimalBuiltinFunc{bf}, false}
+		} else {
+			bf, err := newBaseBuiltinFuncWithTp(args, ctx, retTp, tpInt)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			sig = &builtinUnaryMinusIntSig{baseIntBuiltinFunc{bf}}
+		}
+	case types.ClassDecimal:
+		bf, err := newBaseBuiltinFuncWithTp(args, ctx, retTp, tpDecimal)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		sig = &builtinUnaryMinusDecimalSig{baseDecimalBuiltinFunc{bf}, false}
+	case types.ClassReal:
+		bf, err := newBaseBuiltinFuncWithTp(args, ctx, retTp, tpReal)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		sig = &builtinUnaryMinusRealSig{baseRealBuiltinFunc{bf}}
+	case types.ClassString:
+		tp := argExpr.GetType().Tp
+		if types.IsTypeTime(tp) {
+			bf, err := newBaseBuiltinFuncWithTp(args, ctx, retTp, tpDecimal)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			sig = &builtinUnaryMinusDecimalSig{baseDecimalBuiltinFunc{bf}, false}
+		} else if tp == mysql.TypeDuration {
+			bf, err := newBaseBuiltinFuncWithTp(args, ctx, retTp, tpDecimal)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			sig = &builtinUnaryMinusDecimalSig{baseDecimalBuiltinFunc{bf}, false}
+		} else {
+			bf, err := newBaseBuiltinFuncWithTp(args, ctx, retTp, tpReal)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			sig = &builtinUnaryMinusRealSig{baseRealBuiltinFunc{bf}}
+		}
+	}
+
+	return sig.setSelf(sig), errors.Trace(b.verifyArgs(args))
```

Comment:
we may put b.verifyArgs at the beginning of this func to check the count of args.